Document local testing (#1880)
openshift/openshift-azure: docs/testing.md

```
Upstream issue: https://github.com/openshift/ci-operator-prowgen/issues/79 

## Verify 

This make target runs a series of checks to **verify** if it complies with standard Golang conventions and practices.  This can be run **locally** with the command `make verify`.  When creating a pull-request, this test is part of the acceptance criteria. 

## Unit test

Packages under `pkg` directory contain their individual unit tests.

To run all unit tests **locally**, execute: `make unit`

> _Note: This will **generate** bindata.go files when using this Makefile target as well as testinsights._

Reviewer: Hm. No changes to `bindata.go` should appear if you do not change files we including in bindata.

For example, if you checkout the repo and run `make unit` or any other command that depends on `make generate` you should not see any new changes.

Reviewer: You probably see it because you are still on pinned go-bindata version. Try update switch to the latest go-bindata.

Reviewer: Sorry, I probably just misunderstood the sentence. Looks good to me.

/lgtm

```
Upstream issue: https://github.com/openshift/ci-operator-prowgen/issues/79 

## Verify 

> This repository runs multiple checks to **verify** if it complies with standard go conventions.  The scripts from the target run through a series of checks to ensure the repository remains compliant with standard go practices. This can be ran **locally** with the command `make verify`.  When creating a pull-request, this tests is part of the acceptance criteria. 

Reviewer 2: ```suggestion
This repository runs a series of checks to **verify** if it complies with standard Golang conventions and practices.  This can be run **locally** with the command `make verify`.  When creating a pull-request, this test is part of the acceptance criteria. 
```

> This repository runs multiple checks to **verify** if it complies with standard go conventions.  The scripts from the target run through a series of checks to ensure the repository remains compliant with standard go practices. This can be ran **locally** with the command `make verify`.  When creating a pull-request, this tests is part of the acceptance criteria. 

Reviewer 3: s/repository/make target/

```makefile
verify:
    ./hack/verify/validate-generated.sh

>     go vet ./...

Reviewer 2: Should be `go get`?

We might not want to include this snippet here since it will change, perhaps we can instead link to the target in the Makefile?

Author: I think it should be `go vet` because it's checking the constructs of the go code written in the directories below.  It will fail if the code written fails to comply with the go vet constructs

Reviewer 2: Oh, okay. I didn't actually read the Makefile, I just thought this was a typo at a glance :) Second half of my comment still applies I think.

Reviewer 3: agree with "We might not want to include this snippet here since it will change" - just remove the snippet

    ./hack/verify/validate-code-format.sh
    ./hack/verify/validate-util.sh
    ./hack/verify/validate-codecov.sh
    go run ./hack/validate-imports/validate-imports.go   cmd hack pkg test
    ./hack/verify/validate-sec.sh
```

## Unit test

Packages under `pkg` directory contain their individual unit tests.

To run all unit tests **locally**, execute: `make unit`

```makefile
unit: generate testinsights
  go test ./... -coverprofile=coverage.out -covermode=atomic -json | ./testinsights

> ```

Reviewer 3: again no need for the snippet

_Note: This will **generate** bindata.go files when using this Makefile target as well as testinsights._

or 

at the root of the repository `openshift-azure/` 

> ```go 

Reviewer 2: This is technically a bash command, not go code, right?

go test ./... -v 
```

To run a single package's tests:
```go
go test ./pkg/util/tls
```

To run a subset of tests in a package:
```go

> go test -run TestFoo ./pkg/util/tls

Reviewer 3: does anyone run this on the cli? - yikes I hope not, it is so much easier to run it within your editor..

```

## OpenShift E2E tests

The project has its own end-to-end testing test suite. You can find it under:
`test/e2e/`. It uses [ginko](https://github.com/onsi/ginkgo) and [gomega](https://github.com/onsi/gomega).

To run these tests you will need to have OpenShift cluster running.

> _Note: To have a OpenShift cluster running refer to the following [documentation](../README.md). To create a cluster use `make create` at the root of this repository with the prerequisites from the documentation above._

Reviewer 2: ```suggestion
_Note: To create a cluster use `make create` at the root of this repository with the prerequisites from the the following [documentation](../README.md)._
```

> The cluster running that you are trying to test **must** have the required files and environment variables described in [e2e requirements](e2e/requirements.md).

Reviewer 4: The running cluster ...
